BUG: date_range issue with sub-second granularity

gabrielreid opened the pull request:
Improves (but doesn't completely resolve) #24110, to avoid rounding
issues with sub-second granularity timestamps when creating a
date range.

- [x] closes #24110
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
BUG: date_range issue with millisecond resolution

Fixes #24110, by avoid floating-point rounding issues with
millisecond resolution or higher timestamps when creating a date
range.

pep8speaks commented:
Hello @gabrielreid! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/arrays/datetimes.py`](https://github.com/gabrielreid/pandas/blob/d5d936fa007c582440c7d91631a038fa9cc90766/pandas/core/arrays/datetimes.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/datetimes/test_date_range.py`](https://github.com/gabrielreid/pandas/blob/d5d936fa007c582440c7d91631a038fa9cc90766/pandas/tests/indexes/datetimes/test_date_range.py) !

mroeschke commented on pandas/core/arrays/datetimes.py:
```diff
@@ -307,7 +307,12 @@ def _generate_range(cls, start, end, periods, freq, tz=None,
+            # pass to np.linspace as much as possible
+            arr = np.linspace(
+                0, end.value - start.value,
+                periods).astype('int64') + start.value
```
You can just pass in `int64` into the dtype argument of `linspace` instead of `astype`ing here.

jbrockmendel replied:
I’ve been wrestling with something similar (#24123).  Does this get fixed if you cast “endpoint” to int64 in generate_range_overflow_safe?

gabrielreid replied:
@jbrockmendel generate_range_overflow_save isn't getting called in this situation, because no frequency has been supplied (which I assume is why np.linspace is used).

mroeschke reviewed (changes requested):
Looks pretty good overall. Also needs a whatsnew entry (v0.24.0).

You mentioned that this fix doesn't solve all resolutions. Do you mind adding another test where this still fails, and xfail that test. We can create a followup issue to address that, but overall I think this is a net improvement.

gabrielreid commented:
Updated this PR to include the whatsnew entry.

I also tried to write a test that demonstrates a failing scenario, but it appears that it's not possible to create Timestamp objects that are far enough in the future with sub-second granularity to trigger this issue, which means that this fix seems to cover all scenarios. I've updated the commit message to reflect this as well.

mroeschke commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1305,6 +1305,7 @@ Datetimelike
 - Bug in :class:`DatetimeIndex` where calling ``np.array(dtindex, dtype=object)`` would incorrectly return an array of ``long`` objects (:issue:`23524`)
 - Bug in :class:`Index` where passing a timezone-aware :class:`DatetimeIndex` and `dtype=object` would incorrectly raise a ``ValueError`` (:issue:`23524`)
 - Bug in :class:`Index` where calling ``np.array(dtindex, dtype=object)`` on a timezone-naive :class:`DatetimeIndex` would return an array of ``datetime`` objects instead of :class:`Timestamp` objects, potentially losing nanosecond portions of the timestamps (:issue:`23524`)
+- Bug in :func:`date_range` where using dates with sub-second granularity could return incorrect values or the wrong number of values in the index (:issue:`24110`)
```
Instead of `sub-second granularity` could you say `microsecond resolution or higher`?

After that LGTM

jreback commented on pandas/tests/indexes/datetimes/test_date_range.py:
```diff
@@ -769,3 +769,11 @@ def test_all_custom_freq(self, freq):
         with pytest.raises(ValueError, match=msg.format(freq=bad_freq)):
             bdate_range(START, END, freq=bad_freq)
+
+    def test_range_with_millisecond_granularity(self):
```
can you add all 3 of the cases from the OP via parameterizeation here

jreback commented:
lgtm. @mroeschke over to you.
